dot/services_test.go

```diff
@@ -0,0 +1,656 @@
+	require.NoError(t, err)
+
+	stateSrvc, err := nodeInstance.createStateService(cfg)
+	require.NoError(t, err)
+
+	tests := []struct {
+		name           string
+		service        *state.Service
+		expectedBaseDB *state.BaseState
+		err            error
+	}{
+		{
+			name:           "working example",
+			service:        stateSrvc,
+			expectedBaseDB: stateSrvc.Base,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := nodeInstance.createRuntimeStorage(tt.service)
+			assert.ErrorIs(t, err, tt.err)
+			assert.Equal(t, tt.expectedBaseDB, got.BaseDB)
+			assert.NotNil(t, got.LocalStorage)
+			assert.NotNil(t, got.PersistentStorage)
+			assert.IsType(t, &runtime.NodeStorage{}, got)
+		})
+	}
+}
+
+func Test_createSystemService(t *testing.T) {
+	cfg := NewTestConfig(t)
+
+	genFile := newTestGenesisRawFile(t, cfg)
+
+	cfg.Init.Genesis = genFile
+
+	nodeInstance := nodeBuilder{}
+	err := nodeInstance.initNode(cfg)
+	require.NoError(t, err)
+
+	stateSrvc, err := nodeInstance.createStateService(cfg)
+	require.NoError(t, err)
+
+	type args struct {
+		cfg     *types.SystemInfo
+		service *state.Service
+	}
+	tests := []struct {
+		name      string
+		args      args
+		expectNil bool
+		err       error
+	}{
+		{
+			name: "working example",
+			args: args{
+				service: stateSrvc,
+			},
+			expectNil: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := nodeInstance.createSystemService(tt.args.cfg, tt.args.service)
+			assert.ErrorIs(t, err, tt.err)
+
+			// TODO: change this check to assert.Equal after state.Service interface is implemented.
+			if tt.expectNil {
+				assert.Nil(t, got)
+			} else {
+				assert.NotNil(t, got)
+				assert.IsType(t, &system.Service{}, got)
+			}
+		})
+	}
+}
+
+func Test_newInMemoryDB(t *testing.T) {
+	tests := []struct {
+		name      string
+		path      string
+		expectNil bool
+		err       error
+	}{
+		{
+			name:      "working example",
+			path:      "test_data",
+			expectNil: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := newInMemoryDB(tt.path)
+			assert.ErrorIs(t, err, tt.err)
+
+			if tt.expectNil {
+				assert.Nil(t, got)
+			} else {
+				assert.NotNil(t, got)
+			}
+		})
+	}
+}
+
+//go:generate mockgen -destination=mock_babe_builder_test.go -package $GOPACKAGE github.com/ChainSafe/gossamer/lib/babe ServiceIFace,ServiceBuilder
+
+func Test_nodeBuilder_createBABEService(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockBabeBuilder := NewMockServiceBuilder(ctrl)
+	mockBabeIFace := NewMockServiceIFace(ctrl)
+	mockBabeBuilder.EXPECT().NewServiceIFace(gomock.AssignableToTypeOf(&babe.ServiceConfig{})).DoAndReturn(func(cfg *babe.
+		ServiceConfig) (babe.ServiceIFace, error) {
+		if reflect.ValueOf(cfg.BlockState).Kind() == reflect.Ptr && reflect.ValueOf(cfg.BlockState).IsNil() {
+			return nil, babe.ErrNilBlockState
+		}
+		return mockBabeIFace, nil
+	}).Times(2)
```

Comment:
define your mocks in the subtest instead (see [dis](https://gist.github.com/qdm12/99358e2529a5887ae13a936a7f1c8ae5))

Reply:
Updated.